core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -105,149 +105,59 @@ sealed trait Channel[F[_], A] {
-      else open
-
-    (F.ref(open), F.deferred[Unit]).mapN { (state, closedGate) =>
+  private[this] def impl[F[_]: Concurrent, A](q: Queue[F, A]): F[Channel[F, A]] =
+    Concurrent[F].deferred[Unit].map { closedR =>
       new Channel[F, A] {
 
-        def sendAll: Pipe[F, A, Nothing] = { in =>
-          (in ++ Stream.exec(close.void))
-            .evalMap(send)
-            .takeWhile(_.isRight)
-            .drain
-        }
+        def sendAll: Pipe[F, A, Nothing] =
+          _.evalMapChunk(send(_)).onComplete(Stream.exec(close.void)).drain
 
-        def send(a: A) =
-          F.deferred[Unit].flatMap { producer =>
-            F.uncancelable { poll =>
-              state.modify {
-                case s @ State(_, _, _, _, closed @ true) =>
-                  (s, Channel.closed.pure[F])
-
-                case State(values, size, waiting, producers, closed @ false) =>
-                  if (size < capacity)
-                    (
-                      State(a :: values, size + 1, None, producers, false),
-                      notifyStream(waiting)
-                    )
-                  else
-                    (
-                      State(values, size, None, (a, producer) :: producers, false),
-                      notifyStream(waiting) <* waitOnBound(producer, poll)
-                    )
-              }.flatten
-            }
+        // doesn't interrupt taking in progress
+        def close: F[Either[Channel.Closed, Unit]] =
+          closedR.complete(()).map {
+            case false => Left(Channel.Closed)
+            case true  => Right(())
           }
 
-        def close =
-          state
-            .modify {
-              case s @ State(_, _, _, _, closed @ true) =>
-                (s, Channel.closed.pure[F])
-
-              case State(values, size, waiting, producers, closed @ false) =>
-                (
-                  State(values, size, None, producers, true),
-                  notifyStream(waiting) <* signalClosure
-                )
-            }
-            .flatten
-            .uncancelable
-
-        def isClosed = closedGate.tryGet.map(_.isDefined)
-
-        def closed = closedGate.get
+        def isClosed: F[Boolean] = closedR.tryGet.map(_.isDefined)
 
-        def stream = consumeLoop.stream
+        def send(a: A): F[Either[Channel.Closed, Unit]] =
+          isClosed.ifM(Channel.Closed.asLeft[Unit].pure[F], q.offer(a).map(Right(_)))
 
-        def consumeLoop: Pull[F, A, Unit] =
-          Pull.eval {
-            F.deferred[Unit].flatMap { waiting =>
-              state
-                .modify { state =>
-                  if (shouldEmit(state)) (empty(state.closed), state)
-                  else (state.copy(waiting = waiting.some), state)
+        def stream: Stream[F, A] = {
+          val takeN: F[Chunk[A]] =
+            q.tryTakeN(None).flatMap {
+              case None =>
+                val fallback = Spawn[F].race(q.take, closedR.get).map {
```

Comment:
I don't think this is correct, I think you can lose elements here on a race:
producer `A` successfully enqueues, producer B closes, and now the two notifications for "is closed" and "there are new elements" are racing, if "is closed arrives first", elements are lost

Reply:
Ah! Yes this is indeed an issue. Will ponder…

Reply:
Thought about it some more. This is the adjacent critical regions problem in a different disguise, and the solution is the same: the first `flatMap` is uninterruptible.

`take` on all `Queue` implementations must have the invariant that, if it is canceled, it either preserves the contents of the queue *or* fully produces the `A` it took. Thus, we should be able to `flatMap` with an `uncancelable` region which tosses the `A` somewhere that we can get it. Or, more likely, we use something a bit more powerful than `race`.

Reply:
I think I fixed it. Interestingly, this appears to be the very first (valid) use of `racePair` I've ever seen.

Reply:
> his appears to be the very first (valid) use of racePair I've ever seen.

https://github.com/SystemFw/upperbound/blob/466aacdabd3297aa86221f7adae49627f58fdf7d/modules/core/src/main/scala/internal/Task.scala#L44 ;)

I'll ponder on the implementation later

Reply:
Aha! Your linked use-case is almost identical to this one.